packages/embedder/indexing/retrospectiveDiscussionTopic.ts

```diff
@@ -0,0 +1,267 @@
+import {Selectable} from 'kysely'
+import prettier from 'prettier'
+
+import getRethink, {RethinkSchema} from 'parabol-server/database/rethinkDriver'
+import {DataLoaderWorker} from 'parabol-server/graphql/graphql'
+import getKysely from 'parabol-server/postgres/getKysely'
+import {DB} from 'parabol-server/postgres/pg'
+
+import Comment from 'parabol-server/database/types/Comment'
+import DiscussStage from 'parabol-server/database/types/DiscussStage'
+import MeetingRetrospective, {
+  isMeetingRetrospective
+} from 'parabol-server/database/types/MeetingRetrospective'
+import {RDatum} from 'parabol-server/database/stricterR'
+
+import generateHourIntervals from './generateHourIntervals'
+import {DBInsert} from '../embedder'
+import {upsertEmbeddingsMetaRows} from './embeddingsTablesOps'
+
+export interface EmbeddingsJobQueueRetrospectiveDiscussionTopic
+  extends Omit<DB['EmbeddingsJobQueue'], 'objectType'> {
+  objectType: 'retrospectiveDiscussionTopic'
+}
+
+// Here's a generic reprentation of the text generated here:
+
+// A topic "<DISCUSSION_TOPIC>" was discussed during the meeting "<MEETING_NAME>"
+// that followed the "<TEMPLATE_NAME>" template.
+//
+// <FOR EACH PROMPT>
+// Participants were prompted with, "<PROMPT>: <PROMPT DESCRIPTION>".
+//  - <PARTICIPANT|ANONYMOUS> wrote, "<RESPONSE>"
+//
+// </FOR EACH PROMPT>
+//
+// <IF DISCUSSION>
+// A discussion was held. <DISCUSSION SUMMARY>
+// </END IF DISCUSSION>
+
+const IGNORE_COMMENT_USER_IDS = ['parabolAIUser']
+
+const pg = getKysely()
+
+export async function refreshEmbeddingsMeta(dataLoader: DataLoaderWorker) {
+  const r = await getRethink()
+  const {createdAt: newestMeetingDate} = (await r
```

Comment:
-1 if you use `createdAt` here, but filter by `summarySentAt` below, then meetings that haven't ended by the time the tick comes along will never get handled, unless i misread this?

Reply:
Indeed you're correct. I'll add a comment that this process is merely for building history. The meetings missed by this function (summarySentAt is `null`) will get added to the meta when the meetings complete by adding a hook to the main app.

Reply:
Help me understand further. If we put a hook in the main app that adds a row to the metadata table, that would make `newestMetadata == Date.now()`, which would then be used as the `startDateTime` so no more history would get processed, right?

Reply:
whoops, forgot to un-resolve this issue. this is probably the biggest question mark i've got with the PR. we can merge it, but i wouldn't run the embedder until we get this one figured out otherwise it won't know where to start

Reply:
@mattkrick Oh wow. Good call. I just got what you were inquiring about.

I can think of only one good option. In essence we need a high-water mark for historically items. I think it might look like this:

1. **Establish high-water mark**: we create a table called `EmbeddingsWatermark` or `EmbeddingsHistory` or something

```
id                  PRIMARY KEY
objectType          EmbeddingsObjectTypeEnum
newest              TIMESTAMP
```

2. **Process history using this watermark**: only the embedder process (and the one that has won the election at that) adjusts `newest` after adding each batch to `EmbeddingsMetadata`. If objects are added directly to Metadata directly, this field is not updated.

What do you think?

Reply:
(I also think we should handle this as a separate PR, but let's get the issue added with clear criteria before merging)

Reply:
agreed, we need a watermark. if the operation takes more than a couple minutes, we could store that in a table. if the operation is quick, we can just tell it the date range to process. Either way, we need something that supports processing a particular date range in the case of a bug. Quick pseudo code 

```ts
// redis CLI
PUBLISH buildObjects "{start: '1970-01-01', end: '2024-02-22', objectType: 'discussionTopic'}"

// embedder.ts
const subscriber = new RedistInstance('sub')
subscriber.subscribe('buildObjects')
subscriber.on('message', _channel, message) => {
    addMetadata(JSON.parse(message))
}

const addMetadata = async ({start: Date, end?: Date, objectType?: ObjectType}) => {
  if (objectType === 'discussionTopic') {
    addMetaDataForDiscussions(start, end)
  }
}

const addMetadataForDiscussions(start: Date, end?: Date) {
  let curStart = start
  for (i = 0; i <1e6; i++) {
    const items = r.table('NewMeeting').between(curStart, end, {index: 'endedAt'}).limit(100)
    const failedToInsertDueToUniqueConstraintViolation = await processItems(items)
    if (items.length < 100 || failedToInsertDueToUniqueConstraintViolation) break
    curStart = items[items.length - 1].endedAt
  }
}
```

Reply:
I like breaking up the function this way.

We're always going to need to keep the watermark of the newest historical range in storage, regardless; as you pointed we need  a way to create the time ranges when the embedder is chewing through history. I'll capture both of these things in an issue.

Reply:
Added issue here: https://github.com/ParabolInc/parabol/issues/9451